### DISCUSSION TITLE
postgres:// scheme added to bucket address that already has scheme

### DISCUSSION THREAD

**Discussion by mark-rushakoff** [juicedata/juicefs#2733] [Ideas]
**What happened**:

I am using juicefs-csi-driver with Nomad, with both meta and data stores in postgres. Adding a volume was failing with a cryptic error:

> `lookup postgres on <my dns server>:53: server misbehaving`

**What you expected to happen**:

Volume created correctly.

**How to reproduce it (as minimally and precisely as possible)**:

I was using this volume definition for Nomad:

```hcl
id = "demo-jfs"
name = "demo-jfs"

type = "csi"
plugin_id = "juicefs"

capacity_min = "1MiB"
capacity_max = "10MiB"

capability {
  access_mode     = "single-node-writer"
  attachment_mode = "file-system"
}

secrets {
  name = "timestamp-demo-jfs"
  metaurl = "postgres://juicefs:juicefs@sql-server.node.consul/juicefs_meta"
  bucket = "postgres://juicefs:juicefs@sql-server.node.consul/juicefs_data"
  storage = "postgres"
}
```

I later found out that it was incorrect to have the `postgres://` scheme on the secrets.bucket value, but I had to spend a lot of time debugging before I learned that. I eventually found that 

https://github.com/juicedata/juicefs/blob/54279914e423e0c78852e8166937c3ebf54c645f/pkg/object/sql.go#L163

could simply check `strings.HasPrefix(uri, "postgres://")` before adding that prefix to the `uri`.

Alternatively, the volume could have been rejected for already containing a scheme with `storage=postgres`. Either way would have saved me quite a bit of time.

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: `docker.io/juicedata/juicefs-csi-driver:v0.17.0`
- Cloud provider or hardware configuration running JuiceFS: Locally with Nomad

**Comment by zhijian-pro** [comment]
I think your bucket format is wrong, the format of postgres as an object store is different than when it is a meta.
as meta:
```
juicefs format \
    --storage s3 \
    ... \
    "postgres://user:mypassword@192.168.1.6:5432/juicefs" \
    pics
```
as storage:
```
juicefs format \
    --storage postgres \
    --bucket <host>:<port>/<db>[?parameters] \
    --access-key <username> \
    --secret-key <password> \
    ... \
    myjfs
```

You can refer to this [link](https://juicefs.com/docs/community/how_to_setup_object_storage/#postgresql)

**Comment by mark-rushakoff** [comment]
> I think your bucket format is wrong

Yes, I eventually figured that out:

> I later found out that it was incorrect to have the postgres:// scheme on the secrets.bucket value

It took me several hours to figure out that `postgres://` was required for meta but unallowed for storage.

I filed an issue because of the poor user experience it took to realize that the storage url must not contain a scheme.

My ask in the issue was either:

- Allow `postgres://` in the storage URL when the driver is `postgres`
- Or detect that the URL contains a scheme (`postgres://` or `http://`, etc.), and reject the configuration

This was especially confusing because the Nomad example at https://juicefs.com/docs/csi/csi-in-nomad/#community-edition shows `bucket="http://172.16.254.29:9000/minio/test"`, which makes it look like the scheme is required.

**Comment by zhijian-pro** [comment]
To improve the ease of use of the product, we will ignore the bucket url schema. https://github.com/juicedata/juicefs/issues/2750
